#### Import Packages

In [70]:
# dataset retrieved from https://archive.ics.uci.edu/dataset/352/online+retail
import pandas as pd
import numpy as np
import plotly.express as px

ModuleNotFoundError: No module named 'plotly'

#### Import Data

In [30]:
file_path = "data/Online_Retail.csv"

df = pd.read_csv(file_path,
                 header= 0,
                #  nrows= 5,
                 parse_dates= ["InvoiceDate"]
                 )

print(df.info())


/var/folders/xz/hvt4xjsx52d4vvbdqkszth7h0000gn/T/ipykernel_5613/2940126086.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(file_path,


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB
None


#### Account for null values
- 1454 values in description (dropped)
- 135,080 values in CustomerID, likely guests

540,455 rows left after dropping NA values in description

In [54]:
print("Number of NA values in each column:")
print(df.isna().sum())

total_na = df.isna().sum().sum()
print(f"Total number of NA values in the DataFrame: {total_na}")

# drop rows without descriptions
df = df.dropna(subset=['Description'])

print(df.isna().sum())
# print(len(df))

Number of NA values in each column:
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133626
Country             0
dtype: int64
Total number of NA values in the DataFrame: 133626
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133626
Country             0
dtype: int64


#### Split Dataset
- valid_sales (531167 rows)
- cancellations (9288 rows)


In [62]:
# sort valid sales from cancellations based on the starting 'C'
valid_sales = df[~df['InvoiceNo'].str.startswith('C')] # ~ is the bitwise NOT operator
cancellations = df[df['InvoiceNo'].str.startswith('C')]

# print(len(valid_sales))
# print(len(cancellations))

#### Valid Sales

In [55]:
# print(cancellations[cancellations['StockCode'].str.len() < 5])
# print(valid_sales[valid_sales['StockCode'].str.len() < 5])

#### Cleaning valid_sales
- Items with quantities of 0 or less removed / 474 removed
- Items with StockCode describing postage e.g. POST, M etc. removed / 2,312 removed
- Items with invalid price (£0 or less) removed / 578 removed

527,803 rows left

In [67]:
# print(valid_sales.info())
print(len(valid_sales), "rows before cleaning") # 531,167 rows initially
valid_sales = valid_sales[valid_sales['Quantity'] >= 1] # keep valid quantities / removed 474 rows
# print(len(valid_sales)) # 530,693 rows left
valid_sales = valid_sales[valid_sales['StockCode'].str.len() >= 5] # keep stockcodes for valid items / removed 2312 rows
# print(len(valid_sales)) # 528,381 rows left
valid_sales = valid_sales[valid_sales['UnitPrice'] > 0] # remove any items with price £0 or less / 578 removed
# print(len(valid_sales)) # 527,803 rows left

print(len(valid_sales), "rows after cleaning") # 527,803 rows left

528381 rows before cleaning
527803 rows after cleaning


##### Sales Analysis
Revenue Analysis (By Month)

In [ ]:
# New column to reflect revenue earned from each sale
valid_sales["Revenue"] = valid_sales['Quantity'] * valid_sales['UnitPrice']

# New column YearMonth to separate sales by monthly buckets
valid_sales['YearMonth'] = valid_sales['InvoiceDate'].dt.to_period('M') ## YearMonth in format YYYY-MM

print(valid_sales.head())

monthly_sales = valid_sales.groupby()

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Revenue  \
0 2010-01-12 08:26:00       2.55     17850.0  United Kingdom    15.30   
1 2010-01-12 08:26:00       3.39     17850.0  United Kingdom    20.34   
2 2010-01-12 08:26:00       2.75     17850.0  United Kingdom    22.00   
3 2010-01-12 08:26:00       3.39     17850.0  United Kingdom    20.34   
4 2010-01-12 08:26:00       3.39     17850.0  United Kingdom    20.34   

  YearMonth  
0   2010-01  
1   2010-01  
2   2010-01  
3   2010-01  
4   2010-01  
